sql: honour FORCE option for RLS-enabled tables

spilchen opened the pull request:
This change implements the logic for the FORCE ROW LEVEL SECURITY option on tables with row-level security (RLS) enabled.

By default, table owners are exempt from RLS policies. However, when the FORCE option is set, policies apply to table owners as well. This behavior can be configured via ALTER TABLE ... FORCE ROW LEVEL SECURITY, which was previously introduced.

With this change, when building the plan we now correctly checks the FORCE setting when evaluating RLS policies. If FORCE is enabled, policies apply even when the current user is the table owner.

Additionally, updates to the test catalog allow modifying table ownership (ALTER TABLE ... OWNER) and toggling the FORCE ROW LEVEL SECURITY option, enabling tests to verify behavior changes when FORCE is on or off.

Epic: CRDB-45203
Release note: None
Closes: #138844

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142332)
<!-- Reviewable:end -->

fqazi reviewed (approved):
Great work, just one minor nit <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>:

> Reviewed 18 of 18 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142332#-:-OL-NFmV-1YnPCV0PmAG:bvjt8xr)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/opt/row_level_security.go` line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142332#-OL-M7lU1os-cbheRd3w:-OL-M7lU1os-cbheRd3x:b-3ms18l) ([raw file](https://github.com/cockroachdb/cockroach/blob/27350808de306504fd979a3908f86f9702ca2d76/pkg/sql/opt/row_level_security.go#L40)):*
> ```Go
> 	// NoForceExempt is a map that tracks if any of the tables had their policies
> 	// exempt because they were the table owner and force RLS wasn't set.
> 	NoForceExempt map[TableID]bool
> ```

Nit: would it be cleaner to just have single map with a struct?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142332)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi and @michae2)

___
*[`pkg/sql/opt/row_level_security.go` line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142332#-OL-M7lU1os-cbheRd3w:-OL-fy2k7bfv6I_vEIE7:bmiskcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/27350808de306504fd979a3908f86f9702ca2d76/pkg/sql/opt/row_level_security.go#L40)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: would it be cleaner to just have single map with a struct?
</blockquote></details>

Yes, good idea.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+

spilchen commented:
bors r-

spilchen committed:
sql: honour FORCE option for RLS-enabled tables

This change implements the logic for the FORCE ROW LEVEL SECURITY option
on tables with row-level security (RLS) enabled.

By default, table owners are exempt from RLS policies. However, when the
FORCE option is set, policies apply to table owners as well. This
behavior can be configured via ALTER TABLE ... FORCE ROW LEVEL SECURITY,
which was previously introduced.

With this change, when building the plan we now correctly checks the FORCE
setting when evaluating RLS policies. If FORCE is enabled, policies
apply even when the current user is the table owner.

Additionally, updates to the test catalog allow modifying table
ownership (ALTER TABLE ... OWNER) and toggling the FORCE ROW LEVEL
SECURITY option, enabling tests to verify behavior changes when FORCE is
on or off.

Epic: CRDB-45203
Release note: None
Closes: #138844

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice tests! This looks great.

> Reviewed 10 of 18 files at r1, 1 of 3 files at r2, 7 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142332#-:-OLLFvoZEv62fkJwrrxo:b-iilny5)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @fqazi and @mgartner)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+

spilchen commented:
bors retry
